### ISSUE TITLE
CLOB query halts

### ISSUE THREAD

**Issue by markotikvic** [rana/ora#195]
Hi,
When I try to query data from a CLOB column the driver halts (everything halts).
Queries over other tables with numeric and char data types work without problems, except I get the "maximum number of cursors exceeded" from time to time, but that is survivable. The halting problem, however, isn't.
This is the function:
```go
func fetchJSON(db *ora.Ses, project string) (string, error) {
        stmt, err := db.Prep(`SELECT JSON_CLOB FROM TABLE WHERE PROJECT_NAME` + EqualQuotes(project), ora.S)
        defer stmt.Close()
        if err != nil {
                return "", err
        }
        fmt.Println("prep passes") // this prints

        rset, err := stmt.Qry()
        if err != nil {
                fmt.Println(err) // this does not
                return "", err
        }
        fmt.Println("query passes") // also doesn't print

        var data string
        if rset.Next() {
                data = rset.Row[0].(string)
        }

        return data, nil
}
```
This is a problem with go 1.7/1.8 and ora v4.
Any ideas?

Edit: I should probably mention that this is a huge CLOB I'm trying to fetch. Last setup that this works with is: ora v3 and go 1.8.

**Comment by tgulacsi** [comment]
Strange. We'll have to write a test case to reproduce it.

Try running it, and when halts, hit Ctrl-\ (or send SIGHUP to the process), and inspect/copy here the full stack trace - it'll show where each goroutine is blocked.

And try it with the current master, or bisect where it gone awry.

**Comment by markotikvic** [comment]
Thank you for the fast reply.
Here's the stack after it halts:
```
prep passes
^\SIGQUIT: quit
PC=0x459891 m=0 sigcode=128

goroutine 0 [idle]:
runtime.futex(0xa35270, 0x0, 0x0, 0x0, 0x0, 0xe6a6fd6, 0x0, 0x0, 0x7fff456fde10, 0x41154b, ...)
	/usr/local/go/src/runtime/sys_linux_amd64.s:422 +0x21
runtime.futexsleep(0xa35270, 0x0, 0xffffffffffffffff)
	/usr/local/go/src/runtime/os_linux.go:45 +0x62
runtime.notesleep(0xa35270)
	/usr/local/go/src/runtime/lock_futex.go:145 +0x6b
runtime.stopm()
	/usr/local/go/src/runtime/proc.go:1650 +0xad
runtime.findrunnable(0xc420023300, 0x0)
	/usr/local/go/src/runtime/proc.go:2102 +0x2e4
runtime.schedule()
	/usr/local/go/src/runtime/proc.go:2222 +0x14c
runtime.park_m(0xc4200001a0)
	/usr/local/go/src/runtime/proc.go:2285 +0xab
runtime.mcall(0x7fff456fdfa0)
	/usr/local/go/src/runtime/asm_amd64.s:269 +0x5b

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc42010200c)
	/usr/local/go/src/runtime/sema.go:47 +0x34
sync.(*RWMutex).RLock(0xc420102000)
	/usr/local/go/src/sync/rwmutex.go:43 +0x49
gopkg.in/rana/ora%2ev4.(*defLob).define(0xc4200a2780, 0x1, 0x820070, 0x18, 0xc420102000, 0x0, 0x0)
	/home/marko/workspace/go/src/gopkg.in/rana/ora.v4/defLob.go:50 +0xf8
gopkg.in/rana/ora%2ev4.(*Rset).open(0xc420102000, 0xc42009cea0, 0x2a9c140, 0x0, 0x0)
	/home/marko/workspace/go/src/gopkg.in/rana/ora.v4/rset.go:751 +0x2868
gopkg.in/rana/ora%2ev4.(*Stmt).qryC(0xc42009cea0, 0xa119a0, 0xc420072340, 0x0, 0x0, 0x0, 0xc420102000, 0x0, 0x0)
	/home/marko/workspace/go/src/gopkg.in/rana/ora.v4/stmt.go:446 +0x3c4
gopkg.in/rana/ora%2ev4.(*Stmt).qry(0xc42009cea0, 0x0, 0x0, 0x0, 0xc, 0x0, 0x0)
	/home/marko/workspace/go/src/gopkg.in/rana/ora.v4/stmt.go:383 +0x65
gopkg.in/rana/ora%2ev4.(*Stmt).Qry(0xc42009cea0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc42009cea0)
	/home/marko/workspace/go/src/gopkg.in/rana/ora.v4/stmt.go:378 +0x49
git.to-net.rs/marko.tikvic/webutility.fetchJSON(0xc4200b02c0, 0x84d6eb, 0x7, 0x0, 0x0, 0x0, 0x0)
	/home/marko/workspace/go/src/git.to-net.rs/marko.tikvic/webutility/json_utility.go:170 +0x235
git.to-net.rs/marko.tikvic/webutility.InitTables(0xc4200b02c0, 0x84d6eb, 0x7, 0x0, 0x0)
	/home/marko/workspace/go/src/git.to-net.rs/marko.tikvic/webutility/json_utility.go:146 +0x58
main.main()
	/home/marko/workspace/go/src/git.to-net.rs/marko.tikvic/inis-server/main.go:62 +0x3c2

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2197 +0x1

rax    0xca
rbx    0x0
rcx    0xffffffffffffffff
rdx    0x0
rdi    0xa35270
rsi    0x0
rbp    0x7fff456fdde0
rsp    0x7fff456fdd98
r8     0x0
r9     0x0
r10    0x0
r11    0x286
r12    0x0
r13    0x14
r14    0x0
r15    0xf1
rip    0x459891
rflags 0x286
cs     0x33
fs     0x0
gs     0x0
```
I'll let you know how it goes with the current master.

**Comment by tgulacsi** [comment]
Ok, check the master (fa34ede8c4f1b83b6c6c6565745e8670629a2b5a)
  
  cd $GOPATH/src/gopkg.in/rana/ora.v4
  git fetch origin
  git branch -m master M
  git checkout -b master origin/master

'cause the defLob.go:50 is already commented out :)

**Comment by tgulacsi** [comment]
If this helps, I'll push this as a new version (v4.1.9)

**Comment by markotikvic** [comment]
Did as you said, but still no new commit.
This is what ```go define``` looks like in ```defLob.go:50```
```go
func (def *defLob) define(position int, sqlt C.ub2, gct GoColumnType, rset *Rset) error {
        def.Lock()
        defer def.Unlock()
        def.rset = rset
        def.gct = gct
        def.sqlt = sqlt
        if def.lobs != nil {
                C.free(unsafe.Pointer(&def.lobs[0]))
        }
        rset.RLock()
        fetchLen := rset.fetchLen
        env := rset.stmt.ses.srv.env
        rset.RUnlock()
        def.lobs = (*((*[MaxFetchLen]*C.OCILobLocator)(C.malloc(C.size_t(fetchLen) * C.sof_LobLocatorp))))[:fetchLen]
        if err := def.ociDef.defineByPos(position, unsafe.Pointer(&def.lobs[0]), int(C.sof_LobLocatorp), int(sqlt)); err != nil {
                return err
        }
        prefetchLength := C.boolean(C.TRUE)
        return env.setAttr(unsafe.Pointer(def.ocidef), C.OCI_HTYPE_DEFINE,
                unsafe.Pointer(&prefetchLength), 0, C.OCI_ATTR_LOBPREFETCH_LENGTH)
}
```

**Comment by tgulacsi** [comment]
cd $GOPATH/src/gopkg.in/rana
  rm -rf ora.v4
  git clone https://github.com/rana/ora.git ora.v4

?

**Comment by markotikvic** [comment]
That's it. It's working now.
Thank you.

**Comment by tgulacsi** [comment]
Thanks for the test.
I've published as v4.1.9
